item.lua

```diff
@@ -0,0 +1,300 @@
+-- helper functions that should probably be moved elsewhere
+-----------------------------------------------------------
+
+-- check whether an item is inside a burrow
+local function containsItem(burrow,item)
+    local res = false
+    local x,y,z = dfhack.items.getPosition(item)
+    if x then
+        res = dfhack.burrows.isAssignedTile(burrow, xyz2pos(x,y,z))
+    end
+    return res
+end
+
+-- check whether the item is reachable by one of the citizens
+local function isReachable(item)
+    local reachable = false
+    local x, y, z = dfhack.items.getPosition(item)
+    if x then -- item has a valid position
+        local citizens = dfhack.units.getCitizens(true)
+        for _, unit in pairs(citizens) do
+            if dfhack.maps.canWalkBetween(unit.pos, xyz2pos(x, y, z)) then
+                reachable = true
+                break
+            end
+        end
+    end
+    return reachable
+end
+
+-- fast reachability test for items that requires precomputed walkability groups for
+-- all citizens. Returns false for items w/o valid position (e.g., items in inventories).
+--- @param item item
+--- @param wgroups table<integer,boolean>
+--- @return boolean
+local function fastReachable(item,wgroups)
+    local x, y, z = dfhack.items.getPosition(item)
+    if x then -- item has a valid position
+        local igroup = dfhack.maps.getWalkableGroup(xyz2pos(x, y, z))
+        if wgroups[igroup] then
+            return true
+        else
+            return false
+        end
+    else
+        return false
+    end
+end
+
+--- @return table<integer,boolean>
+local function citizenWalkabilityGroups()
+    local cgroups = {}
+    for _, unit in pairs(dfhack.units.getCitizens(true)) do
+        local wgroup = dfhack.maps.getWalkableGroup(unit.pos)
+        cgroups[wgroup] = true
+    end
+    cgroups[0] = false -- exclude unwarkable tiles
+    return cgroups
+end
+
+-- see: https://discord.com/channels/793331351645323264/807443277132595211/1173979892036141137
+--- @param item item
+local function isMeltable(item)
+    local type = item:getType()
+    local matinfo = dfhack.matinfo.decode(item)
+    return
+        matinfo:matches{metal=true} and
+        not item.flags.artifact and
+        df.item_type[type] ~= 'CORPSE' and
+        df.item_type[type] ~= 'CORPSEPIECE' and
+        df.item_type[type] ~= 'REMAINS' and
+        df.item_type[type] ~= 'FISH' and
+        df.item_type[type] ~= 'FISH_RAW' and
+        df.item_type[type] ~= 'VERMIN' and
+        df.item_type[type] ~= 'PET' and
+        df.item_type[type] ~= 'FOOD' and
+        df.item_type[type] ~= 'EGG'
+end
+
+local function addCondition(tab, cond)
+    if cond then
+        table.insert(tab, cond)
+    end
+end
+-----------------------------------------------------------------------
+-- external API: helpers to assemble filters and `item.act` to execute.
+-----------------------------------------------------------------------
+item = {}
```

Comment:
public APIs are normally expressed via global functions. They will already be namespaced from the local variable that holds the `reqscript` return value. With your API held inside the `item` table, clients will call your code like this:

```lua
local itemtools = reqscript('item`)

local conditions = {}
itemtools.item.condition_type(conditions, 'amulet')
itemtools.item.act('melt', conditions, {})
```

This may be cleaner if there were two top level public functions:

```lua
function add_condition(conditions, type, match)
    ...
end

function apply(action, conditions, options)
    options = options or {}
    ...
end

Reply:
Indeed, looking back at this I'm not happy with the API. Part of this is due to me being upset that, while functional, Lua doesn't handle partial applications very well. However, that was before I saw `dfhack.curry`. With that, factoring things as
```lua
local function condition_material (material, item) return ...    -- definition
table.insert(conditions, curry(condition_material,material))     -- argument parsing / `add_condition`
```
looks more reasonable.

What do you imagine for the `type` argument of the global `add_condition` function. Strings or the fields of something like a "pseudo enum":
```
Condition = { material = {}, description = {}, ... }
```

Reply:
I ultimately decided not to bunch all the filter-generating functions into one. The different functions take, in addition to the obligatory table, 0-2 arguments of different types. So I don't think this would make a lot of sense.

Reply:
the new refactored API looks good to me